src/isa/riscv64/instr/rva/amo.c

```diff
@@ -90,13 +99,16 @@ def_rtl(amo_slow_path, rtlreg_t *dest, const rtlreg_t *src1, const rtlreg_t *src
 
 #ifdef CONFIG_RV_ZACAS
   if (funct5 == 0b00101) { // amocas
-    width = BITS(s->isa.instr.r.funct3, 2, 2) == 0 ? width : 16;
-    int rd = s->isa.instr.r.rd;
-    int rs2 = s->isa.instr.r.rs2;
-    if (width == 16 && ((rd % 2 == 1) || (rs2 % 2 == 1))) {
-      longjmp_exception(EX_II);
-    }
     cpu.amo = true;
+    // check store behavior before actually load or store
+    uint64_t paddr = *dsrc1;
+    if (isa_mmu_check(*dsrc1, width, MEM_TYPE_WRITE) == MMU_TRANSLATE) {
+      paddr = isa_mmu_translate(*dsrc1, width, MEM_TYPE_WRITE);
+    }
+    if (!in_pmem(paddr) || !isa_pmp_check_permission(paddr, width, MEM_TYPE_WRITE, cpu.mode)) {
```

Comment:
cpu.mode should be MPRV?mstatus.MPP:cpu.mode

Reply:
The memory access mode has been obtained in `isa_pmp_check_permission`.

https://github.com/OpenXiangShan/NEMU/blob/master/src/isa/riscv64/system/mmu.c#L1017
>   mode = (out_mode == MODE_M) ? (mstatus->mprv && !ifetch ? mstatus->mpp : cpu.mode) : out_mode;